## Hands-On Data Preprocessing in Python
Learn how to effectively prepare data for successful data analytics
    
    AUTHOR: Dr. Roy Jafari 

### Chapter 3: Data

##### Excercise 1
1)	From 5 colleagues or classmates ask to provide a definition for the term data. 

    a)	Report these definitions and indicate the similarity among them. 
    b)	In your own words, define the all-encompassing definition of data put forth in this chapter.
    c)	Indicate the two important aspects of the definition in b.
    d)	Compare the 5 definitions of data from your colleagues with the all-encompassing definitions and indicate  


##### Excercise 2

For this exercise, we are going to use covid_impact_on_airport_traffic.csv. Answer the following questions. This dataset is from Kaggle.com, use this link to see its page: https://www.kaggle.com/terenceshin/covid19s-impact-on-airport-traffic.
The key attribute of this dataset is PercentOfBaseline which shows the ratio of air traffic in the specific day compared to pre-pandemic time (1st Feb to 15th March 2020)

In [1]:
import pandas as pd
pd.read_csv('covid_impact_on_airport_traffic.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'covid_impact_on_airport_traffic.csv'

    a)	What is the best definition of the data object for this dataset?
    b)	Are there any attributes in the data that only have one value? Use .unique() function to check. If there are remove them from the data and update the definition of the data object.
    c)	What type of values do the remaining attributes carry?
    d)	How much statistical information the attribute ‘PercentOfBaseline’ has?


Answer: 

##### Excercise 3    

For this exercise, we are going to use US_Accidents.csv. Answer the following questions. This dataset is from Kaggle.com, use this link to see its page: https://www.kaggle.com/sobhanmoosavi/us-accidents.
This dataset shows all the car accidents in the US from February 2016 to Dec 2020. 



In [2]:
import pandas as pd
pd.read_csv('US_Accidents.csv')


FileNotFoundError: [Errno 2] No such file or directory: 'US_Accidents.csv'

    a)	What is the best definition of the data object for this dataset?
    b)	Are there any attributes in the data that only have one value? Use .unique() function to check. If there are remove them from the data and update the definition of the data object.
    c)	What type of values do the remaining attributes carry?
    d)	How much statistical information the numerical attributes of the dataset carry?
    e)	Compare the statistical information of the numerical attributes and see if any of them are a candidate for data redundancy.

Answer:

##### Excercise 4 

For this exercise, we are going to use fatal-police-shootings-data.csv. There are a lot of debates, discussions, dialogues, and protests happening in the US surrounding police killings. The Washington Post has been collecting data on all fatal police shootings in the US. The dataset available to the government and the public alike has date, age, gender, race, location, and other situational information of these fatal police shootings. You can read more about this data on https://www.washingtonpost.com/graphics/investigations/police-shootings-database/, and you can download the last version of the data from https://github.com/washingtonpost/data-police-shootings

In [3]:
import pandas as pd

pd.read_csv('fatal-police-shootings-data.csv').head()

id                name        date   manner_of_death       armed   age  \
0   3          Tim Elliot  2015-01-02              shot         gun  53.0   
1   4    Lewis Lee Lembke  2015-01-02              shot         gun  47.0   
2   5  John Paul Quintero  2015-01-03  shot and Tasered     unarmed  23.0   
3   8     Matthew Hoffman  2015-01-04              shot  toy weapon  32.0   
4   9   Michael Rodriguez  2015-01-04              shot    nail gun  39.0   

  gender race           city state  signs_of_mental_illness threat_level  \
0      M    A        Shelton    WA                     True       attack   
1      M    W          Aloha    OR                    False       attack   
2      M    H        Wichita    KS                    False        other   
3      M    W  San Francisco    CA                     True       attack   
4      M    H          Evans    CO                    False       attack   

          flee  body_camera  longitude  latitude  is_geocoding_exact  
0  Not fleeing        False   -123.122    47.247                True  
1  Not fleeing        False   -122.892    45.487                True  
2  Not fleeing        False    -97.281    37.695                True  
3  Not fleeing        False   -122.422    37.763                True  
4  Not fleeing        False   -104.692    40.384                True

    a)	What is the best definition of the data object for this dataset?
    b)	Are there any attributes in the data that only have one value? Use .unique() function to check. If there are remove them from the data and update the definition of the data object.
    c)	What type of values do the remaining attributes carry?
    d)	How much statistical information the numerical attributes of the dataset carry?
    e)	Compare the statistical information of the numerical attributes and see if any of them are a candidate for data redundancy.


##### Excercise 5
For this exercise, we will be using electricity_prediction.csv. The screenshot below shows the 5 rows of this dataset and a linear regression model created to predict electricity consumption based on the weekday and daily average temperature. 

In [4]:
import pandas as pd

electricity_df = pd.read_csv('electricity_prediction.csv')
electricity_df.head()

Date  Weekday  Consumption  Average Temperature
0  1/1/2016        4      2581914                   80
1  1/2/2016        5      2663011                   77
2  1/3/2016        6      2725351                   78
3  1/4/2016        0      3092978                   80
4  1/5/2016        1      3231827                   81

In [7]:
from sklearn.linear_model import LinearRegression

X = electricity_df[['Weekday', 'Average Temperature']]
y = electricity_df['Consumption']
X

Weekday  Average Temperature
0          4                   80
1          5                   77
2          6                   78
3          0                   80
4          1                   81
..       ...                  ...
361        1                   81
362        2                   81
363        3                   81
364        4                   81
365        5                   81

[366 rows x 2 columns]

In [8]:

lrm = LinearRegression()
lrm.fit(X, y)

print('intercept ', lrm.intercept_)
print(pd.DataFrame({'Predictor': X.columns, 'coefficient': lrm.coef_}))

intercept  3074181.49501588
             Predictor   coefficient
0              Weekday -55710.145405
1  Average Temperature  -3476.377056


The regression model that is derived from the data is presented below.

    Consumption = 3074181.5- 55710.1 × Weekday-3476.4 ×Average Temperature

What is the fundamental mistake in this analysis? Describe it and provides possible solutions for it. 


Answer:

##### Excercise 6
For this exercise, we will be using adult.csv. we used this dataset extensively in chapter 1. Read the dataset using Padans and call it adult_df.

In [1]:
import pandas as pd

adult_df = pd.read_csv('adult.csv')

    a)	What type of values does the attribute eduction carry?
    b)	Run 'adult_df.education.unique()', study the results, and explain what the code does.
    c)	Based on your understandings, order the output of the code you ran for b).
    d)	Run 'pd.get_dummies(adult_df.education)', study the results, and explain what the code does.
    e)	Run 'adult_df.sort_values(['education-num']).iloc[1:32561:1200]', study the results and explain what the code does.
    f)	Compare your answer to c) and what you learned from e). Was the order you came up with in c) correct?
    g)	Education is an ordinal attribute, translating an ordinal attribute from an analytic perspective to a programming perspective involves choosing between Boolean representation, string representation, and integer representation. Choose which choice has been made for the three following representations of the attribute education.
        - 'adult_df.education'
        - 'pd.get_dummies(adult_df.education)'
        - 'adult_df[‘education’]'
    h)	Either of the choices has some advantages and some disadvantages. Select which programing data representation each statement below describes. 
        - If an ordinal attribute is presented using this programming value representation, no bias or assumptions are added to the data, but algorithms that work with numbers cannot use the attribute. 
        - If an ordinal attribute is presented using this programming value representation, the data can be used by algorithms that only take numbers, but the size of the data becomes bigger and there may be concerns for computational costs.
        - If an ordinal attribute is presented using this programming value representation, there will be no size or computational concerns, but some statistical information that may not be true is assumed and it may create bias.


Answer: